papers/bradley_dice/55_bradley_dice.rst

```diff
@@ -0,0 +1,553 @@
+
+The need to parse these JSON files for indexing and the complexity of modifying them represent the most significant barriers to scaling signac.
+Even in the absence of file modification, reading a large number of files simply to produce a database index becomes prohibitively expensive for large data spaces.
+Although various optimizations have incrementally improved signac's scalability, an alternative means of storing the state point and associated metadata that circumvents the heavy I/O costs of our current approach has the potential to make a much larger impact.
+However, replacing individual JSON files as the primary data source for signac without breaking signac's API would require a generic means for providing the same interface to the underlying index and metadata irrespective of the underlying storage mechanism.
+Once developed, however, such an API would abstract out enough of the internals of signac to enable other generalizations as well, such as making it relatively easy to support alternate (and nearly arbitrary) data space layouts.
+
+The synced collections subpackage of signac represents the culmination of our efforts to expose this functionality, providing a generic framework within which interfaces corresponding to any of Python's built-in types can be easily constructed with arbitrary underlying synchronization protocols.
+For instance, with synced collections it becomes easy to define a new list-like type that automatically saves all its data in a plain-text CSV format.
+However, the flexibility of this new framework extends far beyond that, defining a generic protocol that can be used to provide a dictionary, list, or set-like API to any arbitrary underlying data structure, including other in-memory objects that do not present a similarly Pythonic API.
+
+Summary of Features
+~~~~~~~~~~~~~~~~~~~
+
+We designed synced collections to be flexible, easily extensible, and independent of the rest of signac.
+The central element is the ``SyncedCollection`` class, which defines a new abstract class extending the ``collections.abc.Collection`` from the Python standard library.
+A ``SyncedCollection`` is a ``Collection`` that adds two additional groups of abstract methods that must be implemented by its subclasses.
+One group includes methods for synchronizing with an underlying resource, while the other contains methods for synchronizing with a standard collection of the underlying base type.
+For instance, a ``JSONDict`` would implement the first set of methods to define how to save a dictionary to a JSON file and reload it, while it would implement the second set of methods to define how to convert between a ``JSONDict`` instance and a standard Python dictionary.
+
+Critically, these two sets of functions are orthogonal.
```

Comment:
There's an extra newline before this paragraph begins in the PDF. I'm not sure why, might be related to the figure above it?

Reply:
Seems fixed now.